packages/perf/src/index.ts

```diff
@@ -0,0 +1,172 @@
+}
+
+export interface PerfServiceInit {
+  protocolName?: string
+  maxInboundStreams?: number
+  maxOutboundStreams?: number
+  timeout?: number
+  writeBlockSize?: bigint
+}
+
+export interface PerfServiceComponents {
+  registrar: Registrar
+  connectionManager: ConnectionManager
+}
+
+export class DefaultPerfService implements Startable, PerfService {
+  public readonly protocol: string
+  private readonly components: PerfServiceComponents
+  private started: boolean
+  private readonly databuf: ArrayBuffer
+  private readonly maxInboundStreams: number
+  private readonly maxOutboundStreams: number
+  private readonly timeout: number
+  private readonly writeBlockSize: bigint
+
+  constructor (components: PerfServiceComponents, init: PerfServiceInit) {
+    this.components = components
+    this.started = false
+    this.protocol = init.protocolName ?? PROTOCOL_NAME
+    this.databuf = new ArrayBuffer(Number(init.writeBlockSize ?? WRITE_BLOCK_SIZE))
+    this.maxInboundStreams = init.maxInboundStreams ?? MAX_INBOUND_STREAMS
+    this.maxOutboundStreams = init.maxOutboundStreams ?? MAX_INBOUND_STREAMS
+    this.timeout = init.timeout ?? TIMEOUT
+    this.writeBlockSize = init.writeBlockSize ?? WRITE_BLOCK_SIZE
+  }
+
+  async start (): Promise<void> {
+    await this.components.registrar.handle(this.protocol, (data: IncomingStreamData) => { void this.handleMessage(data) }, {
+      maxInboundStreams: this.maxInboundStreams,
+      maxOutboundStreams: this.maxOutboundStreams
+    })
+    this.started = true
+  }
+
+  async stop (): Promise<void> {
+    await this.components.registrar.unhandle(this.protocol)
+    this.started = false
+  }
+
+  isStarted (): boolean {
+    return this.started
+  }
+
+  async handleMessage (data: IncomingStreamData): Promise<void> {
+    const { stream } = data
+
+    const writeBlockSize = this.writeBlockSize
+
+    let bytesToSendBack: bigint | null = null
+
+    for await (const buf of stream.source) {
+      if (bytesToSendBack === null) {
+        bytesToSendBack = BigInt(buf.getBigUint64(0, false))
+      }
+      // Ingest all the bufs and wait for the read side to close
+    }
+
+    const uint8Buf = new Uint8Array(this.databuf)
+
+    if (bytesToSendBack === null) {
+      throw new Error('bytesToSendBack was null')
+    }
+
+    await stream.sink(async function * () {
+      while (bytesToSendBack > 0n) {
+        let toSend: bigint = writeBlockSize
+        if (toSend > bytesToSendBack) {
+          toSend = bytesToSendBack
+        }
+        bytesToSendBack = bytesToSendBack - toSend
+        yield uint8Buf.subarray(0, Number(toSend))
+      }
+    }())
+  }
+
+  async perf (connection: Connection, sendBytes: bigint, recvBytes: bigint, options: AbortOptions = {}): Promise<void> {
+    log('opening stream on protocol %s to %p', this.protocol, connection.remotePeer)
+
+    const uint8Buf = new Uint8Array(this.databuf)
+
+    const writeBlockSize = this.writeBlockSize
```

Comment:
Why?

Reply:
To reference the writeBlockSize within the `stream.sink` function scope it would need to be removed from the `this` binding of the block scope, likewise the `this.databuf`